### ISSUE TITLE
Submission: UCSCXenaTools

### ISSUE THREAD

**Issue by ShixiangWang** [ropensci/software-review#315]
Submitting Author: ShixiangWang (@ShixiangWang)  
Repository: https://github.com/ShixiangWang/UCSCXenaTools  
Version submitted: 1.2.3   
Editor: @melvidoni 
Reviewer 1: @ChristineStawitz-NOAA
Reviewer 2: @carlganz
Archive: TBD  
Version accepted: TBD   

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: UCSCXenaTools
Title: Download and Explore Datasets from UCSC Xena Data Hubs
Version: 1.2.3
Authors@R: c(person("Shixiang", "Wang", email = "w_shixiang@163.com", role = c("aut", "cre"), 
            comment = c(ORCID = "0000-0001-9855-7357")),
            person("Martin", "Morgan", role="aut"))
Maintainer: Shixiang Wang <w_shixiang@163.com>
URL: https://github.com/ShixiangWang/UCSCXenaTools
BugReports: https://github.com/ShixiangWang/UCSCXenaTools/issues
Description: Download and explore datasets from UCSC Xena data hubs, which are
    a collection of UCSC-hosted public databases such as TCGA, ICGC, TARGET, GTEx, CCLE, and others.
    Databases are normalized so they can be combined, linked, filtered, explored and downloaded.
Depends: R (>= 3.0)
Imports: 
    httr,
    readr,
    methods,
    utils,
    dplyr,
    magrittr,
    jsonlite,
    shinydashboard,
    shiny,
    DT
License: GPL-3
Encoding: UTF-8
LazyData: true
Suggests: 
    knitr,
    rmarkdown,
    prettydoc,
    testthat (>= 2.1.0),
    covr
VignetteBuilder: knitr
RoxygenNote: 6.1.1
Roxygen: list(markdown = TRUE)
```

## Scope 

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [x] data retrieval
	- [ ] data extraction
	- [ ] database access
	- [ ] data munging
	- [ ] data deposition
	- [ ] reproducibility
	- [ ] geospatial data
	- [ ] text analysis
	

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):

The package falls under the **data retrieval category** because it pulls data from [UCSC Xena Hubs](https://xenabrowser.net/datapages/).

-   Who is the target audience and what are scientific applications of this package?  
    -  Target Audience:  cancer and clinical researchers, bioinformaticians
    - Applications: genomic and clinical analyses

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?

None that I am aware of. There is a Python package called [xenaPython](https://github.com/ucscXena/xenaPython) which is the official API for Xena Hubs. However, **UCSCXenaTools** is more powerful and provides more features including download whole dataset and query a small amount of data. 

-   If you made a pre-submission enquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

https://github.com/ropensci/software-review/issues/314

## Technical checks

Confirm each of the following by checking the box.  This package:

- [x] does not violate the Terms of Service of any service it interacts with. 
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme). 
- [x] includes [documentation with examples for all functions](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage using services such as Travis CI, Coveralls and/or CodeCov.

## Publication options

- [ ] Do you intend for this package to go on CRAN?  
This package is already available at CRAN
- [x] Do you wish to automatically submit to the [Journal of Open Source Software](http://joss.theoj.org/)? If so:

<details>
 <summary>JOSS Options</summary>  

  - [ ] The package has an **obvious research application** according to [JOSS's definition](https://joss.readthedocs.io/en/latest/submitting.html#submission-requirements).
    - [ ] The package contains a `paper.md` matching [JOSS's requirements](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain) with a high-level description in the package root or in `inst/`.
    - [ ] The package is deposited in a long-term repository with the DOI: 
    - (*Do not submit your package separately to JOSS*)  
  
</details>

- [ ] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:  

<details>
<summary>MEE Options</summary>  

- [ ] The package is novel and will be of interest to the broad readership of the journal. 
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.github.io/dev_guide/policies.html#code-of-conduct) during the review process and in maintaining my package should it be accepted.

**Comment by melvidoni** [comment]
### Editor checks:

- [X] **Fit**: The package meets criteria for [fit](policies.md#fit) and [overlap](policies.md#fit)
- [X] **Automated tests:** Package has a testing suite and is tested via Travis-CI or another CI service.
- [ ] **License:** The package has a CRAN or OSI accepted license
- [X] **Repository:** The repository link resolves correctly
- [ ] **Archive** (JOSS only, may be post-review): The repository DOI resolves correctly
- [ ] **Version** (JOSS only, may be post-review): Does the release version given match the GitHub release (v1.0.0)?

---

#### Editor comments

Thank you for submitting your package! Please add a LICENCE file to the repo, and fix the following suggestions of `goodpractices::gp()` package:

```
── GP UCSCXenaTools ──────────────────────────────────────────────────────────────────────────

It is good practice to

  ✖ not import packages as a whole, as this can cause name clashes
    between the imported packages. Instead, import only the specific functions you
    need.
────────────────────────────────────────────────────────────────────────────────────────────── 
```



---

Reviewers: TBD
Due date: TBD

**Comment by ShixiangWang** [comment]
Dear editor @melvidoni,

All the comments you mentioned above have been fixed. A [GPL-3 license file](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/LICENSE.md) has been added by `usethis::use_gpl3_license` command. All `import` sentences in [NAMESPACE file](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/NAMESPACE) have been modified to `importFrom`, and [corresponding code](https://github.com/ShixiangWang/UCSCXenaTools/commit/0b374bfbfdaee70d32791609ae76293f8f935e5f) have also been modified. 

Thanks for your comments. Please let me know if there are other issues.

Best,
Shixiang

**Comment by melvidoni** [comment]
Thanks @ShixiangWang! I'm looking for reviewers, which may take a little. I will comment again when I assign them, and let you know about the due date.

**Comment by melvidoni** [comment]
Reviewers @ChristineStawitz-NOAA and @carlganz have been assigned to the package. They'll have until July 22nd to get the reviews in.

**Comment by ShixiangWang** [comment]
Thanks, @melvidoni. Welcome the reviewers👏 👏 👏

**Comment by ChristineStawitz-NOAA** [comment]
Apologies for getting this in at the last possible moment! 

## Package Review

- [x] As the reviewer I confirm that there are no [conflicts of interest](#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README

    **This might be obvious to those who use the USCS data hubs, but the target audience was not clear to me in the README; please add in this information as described above in the README.**

- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README

  **Worked fine on my PC**

- [ ] **Vignette(s)** demonstrating major functionality that runs successfully locally

1. I was able to find the Vignettes on the author's website. However, they are not installed locally either when using `devtools::install_github()` or installing from CRAN. Even if there is a website version that is more up-to-date, it seems like a local version would be useful, particularly for CRAN users who might not find the Github link.
2. I tried some of the intro vignette with the CRAN version and some with the Github one - with the CRAN version, `XenaScan()` is not found. There should be some guidance on minimum versions to run the vignette when describing installation.
3. Intro vignette - For the examples using local files with `XenaPrepare()` it would be better to put some sample .gz files in the `data/` directory and use  rather than referencing `E:/Github` or `./downloads/` which all users won't have. I'm not sure if this is recommended by rOpenSci but in the past I've used `system.file("data", package = "UCSCXenaTools")` to access this directory.
4. API vignette - typo in one of the headers: "Find out the number of idnetifiers in a dataset"

- [x] **Function Documentation:** for all exported functions in R help
- [x] **Examples** for all exported functions in R Help that run successfully locally
- [x] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).
      
  **There is no guidance on how to contribute**

>##### For packages co-submitting to JOSS
>
>- [ ] The package has an **obvious research application** according to [JOSS's definition](http://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](http://joss.theoj.org/about#paper_structure) with:
>
>- [ ] **A short summary** describing the high-level functionality of the software
>- [ ] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [ ] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software been confirmed.
- [x] **Performance:** Any performance claims of the software been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.

 **I get a warning when running `devtools::check()` - "qpdf is needed for size reduction on PDFs"**
 
- [ ] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

1. Argument naming is not consistent - for example, the `fetch_` API functions are in snake_case but a number of other functions are camel-cased with the first letter capitalized (i.e. `XenaScan`) or regular camel-cased (`getTCGAdata`). It would be helpful to adopt a similar casing style even if snake_case doesn't work because of consistency with other tools for these data sets.
2. The Workflow.R and simplify.R files are a bit overloaded - style guidelines suggest not putting all of the key functions in a single file, and it wasn't clear to me where to search for some of the different functions.
3. Throughout the R code, `::` is used to reference functions within the `UCSCXenaTools` package itself as well as `dplyr`, which is imported. It could clean things up to remove these for packages that are explicitly imported. 
4. I think `shiny` and `shinydashboard` could be moved to suggests instead of imports in the `DESCRIPTION`

---

### Review Comments

I'm not in the omics field, but this seems like a useful package for interacting with the UCSC Xena platform. I found the package reliable and stable, but it was a bit hard to follow where the vignettes were located after local install and there were a few vignette issues (see above). There are a couple of code simplifications and consistency improvements that could be made as well, which are detailed above.

**Comment by melvidoni** [comment]
Thank you @ChristineStawitz-NOAA! We are waiting for @carlganz's now, then.

**Comment by carlganz** [comment]
Promise to finish review tomorrow sorry for delay.

**Comment by ShixiangWang** [comment]
@ChristineStawitz-NOAA  Thanks for your all comments. Basically, I fixed all the problems you raised. And I also have some reasons for the non-modified parts.

All replies for problematic points have been put as the following. Please forgive me my English if you find something hard to understand, just ask me again. :)

* [x] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README

  Thanks, this information has been added to [README](https://github.com/ShixiangWang/UCSCXenaTools/)

* [x] **Vignette(s)**

  1. I have created a [local version](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/vignettes/USCSXenaTools.Rmd) describing the basic usage of **UCSCXenaTools**, you can install it from the latest version at GitHub by `remotes::install_github("ShixiangWang/UCSCXenaTools", build_vignettes = TRUE)`
  2. The minimum version has been described at [installation section](https://github.com/ShixiangWang/UCSCXenaTools/blob/88f4e9ff0971ed798cfddec6df8bb964396d841a/vignettes/USCSXenaTools.Rmd#L62)
  3. I have updated the intro vignette and use /tmp directory to describe the message. I don't use local data because it will increase package size dramatically.
  4. The typo has been fixed, thanks :).

* [ ] **Submitting to JOSS**

	I don't know what should I do next to submit the package to JOSS. After reviewing, is this package accepted by JOSS automatically? Or should I write the summary, authors, etc.. now and submit it along with the package to JOSS?
	
* [x] **Test**

I get a warning when running devtools::check() - "qpdf is needed for size reduction on PDFs"

* `qpdf` is required for R package check but not this package, after you install `qpdf` according to your system, the warning will be gone.

* [x] **Packaging**
	1. Argument naming is indeed not consistent.  The naming strategy has a different purpose for now. Functions start with `Xena` are core functions which can be combined by pipe operator `%>%` to operate datasets;  functions like `getTCGAdata`, `avail*` etc. are side effects that I don't want the users to use them but they are useful for some users; functions using snake case basically are API functions for interacting with UCSC Xena database to query/download a part of a dataset. I agree with your consistent idea, but for now, I can't adjust them before I prepare to make an incompatible version. 
	2. Thanks and I have split the `workflow.R` into [individual files](https://github.com/ShixiangWang/UCSCXenaTools/tree/master/R). The functions in`simplify.R` all work for one purpose: select and download TCGA datasets, it will be harder to maintain if I split it up.
	3. The mentioned code I don't like to change because they make me clearly know where the data or functions come from and write less redundant document comments like `@importFrom`.
	4. I have moved `shiny`, `shinydashborad` and `DT` packages to suggests and cleaned [the corresponding code](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/R/shiny.R). This reduces the core dependencies, thanks.

**Comment by carlganz** [comment]
## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- [X] As the reviewer I confirm that there are no [conflicts of interest](#coi) for me to review this work (If you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience in README
- [ ] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [X] **Vignette(s)** demonstrating major functionality that runs successfully locally
- [X] **Function Documentation:** for all exported functions in R help
- [X] **Examples** for all exported functions in R Help that run successfully locally
- [X] **Community guidelines** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

>##### For packages co-submitting to JOSS
>
>- [X] The package has an **obvious research application** according to [JOSS's definition](http://joss.theoj.org/about#submission_requirements)
>
>The package contains a `paper.md` matching [JOSS's requirements](http://joss.theoj.org/about#paper_structure) with:
>
>- [X] **A short summary** describing the high-level functionality of the software
>- [X] **Authors:**  A list of authors with their affiliations
>- [ ] **A statement of need** clearly stating problems the software is designed to solve and its target audience.
>- [X] **References:** with DOIs for all those that have one (e.g. papers, datasets, software).

#### Functionality

- [ ] **Installation:** Installation succeeds as documented.
- [X] **Functionality:** Any functional claims of the software been confirmed.
- [X] **Performance:** Any performance claims of the software been confirmed.
- [X] **Automated tests:** Unit tests cover essential functions of the package
   and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [X] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines

#### Final approval (post-review)

- [ ] **The author has responded to my review and made changes to my satisfaction. I recommend approving this package.**

Estimated hours spent reviewing: 4

- [ ] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

I have very little knowledge of genomics, but I can see clearly that the UCSC Xena platform is valuable to the genomics community, and after thoroughly reviewing the code within your package I can see that working with the Xena platform is highly non-trivial, and that this package encapsulates a great deal of work that succesfully makes the Xena platform much more accessible to R users. Congrats on the wonderful package!

# Statement of Need and JOSS summary

You explain what the Xena platform is, but it would be nice to have a sentence or two explaining why interfacing with the Xena platform is non-trivial, and how this interface significantly improves R user access. 

# Installation

* When I used `remotes::install_github("ShixiangWang/UCSCXenaTools", build_vignettes = TRUE)` I got error about missing `prettydoc` dependency, but I found `remotes::install_github("ShixiangWang/UCSCXenaTools", build_vignettes = TRUE, dependencies = TRUE)` worked fine.

# Small code suggestions

* [Here](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/R/XenaHub-class.R#L215) and a few other places use the discouraged antipattern `1:length(some_vec)` or `1:nrow(some_df)`. `seq_len` is the safer alternative, but I am pretty sure it won't make a difference in this case.

* Another antipattern worth avoiding is iteratively building a data.frame with `rbind` in a loop like [here](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/R/XenaHub-class.R#L244). Section 4 in [this blog post by @noamross](https://www.noamross.net/blog/2014/4/16/vectorization-in-r--why.html) explains why not to do this, but essentially each iteration copys the entire data.frame in memory.

So I would replace this:

```r
  resDF <- data.frame(stringsAsFactors = FALSE)
  for (i in 1:length(XenaList)) {
    hostNames <- names(XenaList)[i]
    cohortNames <- names(XenaList[[i]])
    res <- data.frame(stringsAsFactors = FALSE)

    for (j in 1:length(cohortNames)) {
      oneCohort <- XenaList[[i]][j]
      # The unassigned cohorts have NULL data, remove it
      if (names(oneCohort) != "(unassigned)") {
        resCohort <- data.frame(
          XenaCohorts = names(oneCohort),
          XenaDatasets = as.character(oneCohort[[1]]),
          stringsAsFactors = FALSE
        )
        res <- rbind(res, resCohort)
      }
    }
    res$XenaHosts <- hostNames
    resDF <- rbind(resDF, res)
  }
```
With something like:
```r
  # use lapply to create list of dfs because why introduce purrr dependency
  resDF <- lapply(seq_len(XenaList), function(i) {
    hostNames <- names(XenaList)[i]
    cohortNames <- names(XenaList[[i]])
    res <- data.frame(stringsAsFactors = FALSE)

    for (j in 1:length(cohortNames)) {
      oneCohort <- XenaList[[i]][j]
      # The unassigned cohorts have NULL data, remove it
      if (names(oneCohort) != "(unassigned)") {
        resCohort <- data.frame(
          XenaCohorts = names(oneCohort),
          XenaDatasets = as.character(oneCohort[[1]]),
          stringsAsFactors = FALSE
        )
        res <- rbind(res, resCohort)
      }
    }
    res$XenaHosts <- hostNames
    return(res)
  }) %>%
  # convert list of dfs into single df
  bind_rows
```

* Documentation from dplyr says `dplyr::rbind_list` is deprecated and to use `dplyr::bind_rows`, which I recommend you do [here](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/R/XenaHub-class.R#L308) just in case Hadley removes the function at some point you don't want to be a breaking reverse dependency on CRAN. 

Hope that helps!

**Comment by ShixiangWang** [comment]
@ChristineStawitz-NOAA Sorry for ignoring  `There is no guidance on how to contribute` yesterday, I have added it [here](https://github.com/ShixiangWang/UCSCXenaTools#how-to-contribute).

@carlganz Thanks for your comments :). The modification can be seen from the last [two commits](https://github.com/ShixiangWang/UCSCXenaTools/commits/master).  I also reply your comments point to point as the following. 

#### Statement of Need and JOSS summary

It was a template in repository yesterday. I rewrite it today, you can see new version [here](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/paper/paper.md). I am not good at English and also writing, so any suggestion is welcome :).

#### Installation

**prettydoc** package is in suggests field, so it will not be install automatically without option `dependencies=TRUE`. Thank   you for poiniting it out, I have added this message in [new README](https://github.com/ShixiangWang/UCSCXenaTools)

#### use seq_len

I like `1:length(x)` this format because it is more easy to read. `seq_len` is indeed a safer version for object with length 0. I have changed  corresponding code.

#### use for

I won’t change such code for now. I do like `for` more than `lapply` here because `for` loop is much easier to read and debug. 

For example,

```
  # use lapply to create list of dfs because why introduce purrr dependency
  resDF <- lapply(seq_len(XenaList), function(i) {
    hostNames <- names(XenaList)[i]
    cohortNames <- names(XenaList[[i]])
    res <- data.frame(stringsAsFactors = FALSE)
    ...
```

The code won’t work before modifying all `XenaList` to `i` and removing `[i]`. And I have to debug to make sure the code works.

When I developed this package, I have no idea how to use `purrr`, so this package is not used here.

As Hadley wrote in *Advance in R*, the for loop in R now is very quick. It will not affect the performance.

#### use bind_rows

This is modified.

**Comment by ShixiangWang** [comment]
I am glad to add two reviewers in DESCRIPTION file according to

 > Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

Thanks all your comments @ChristineStawitz-NOAA @carlganz

**Comment by melvidoni** [comment]
Great to see this is moving! Please @ChristineStawitz-NOAA and @carlganz, review the changes. Feel free to go back and forth if you deem it necessary.

**Comment by ChristineStawitz-NOAA** [comment]
**Statement of Need and JOSS paper**
I don't see the bibliography showing up at the end of the paper.md file, though I can see the in-text references. 

I initially did not review this as a JOSS submission, should I be also following their review criteria @melvidoni or will it go through a separate review at JOSS after the rOpenSci one?

I have some long-running code going in other `RStudio` sessions so I will not be able to reinstall the package and check the vignette again until tomorrow - assuming that works I will sign off on the review then as I am satisfied with the changes from what is in this thread and the repo :) Nice work @ShixiangWang ! I am also OK with being credited for the review, thanks.

**Comment by carlganz** [comment]
I am happy with the updates. I've updated review to give full approval. Congrats!

**Comment by ShixiangWang** [comment]
@ChristineStawitz-NOAA the bibliography is shown in [paper.bib](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/paper/paper.bib) file according to [JOSS instruction](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography).

@ChristineStawitz-NOAA @carlganz I have added your names [here](https://github.com/ShixiangWang/UCSCXenaTools/blob/66eba83777e61b5f1a6f2b375b2d2c598cfb8bbf/DESCRIPTION#L8) and [here](https://github.com/ShixiangWang/UCSCXenaTools/blob/master/paper/paper.md#acknowledgements), please take a look and make sure they are right.

**Comment by ChristineStawitz-NOAA** [comment]
Looks good and I am able to successfully run everything in the vignette
after the changes - thanks!

**Comment by melvidoni** [comment]
Approved! Thanks @ShixiangWang  for submitting and @ChristineStawitz-NOAA @carlganz  for your reviews!

To-dos:
- [ ] Transfer the repo to rOpenSci's "ropensci" GitHub organization under "Settings" in your repo.  I have invited you to a team that should allow you to do so.  You'll be made admin once you do.
- [ ] Add the rOpenSci footer to the bottom of your README
```
[![ropensci_footer](https://ropensci.org/public_images/ropensci_footer.png)](https://ropensci.org)
```

- [ ] Fix any links in badges for CI and coverage to point to the ropensci URL. We no longer transfer Appveyor projects to ropensci Appveyor account so after transfer of your repo to rOpenSci's "ropensci" GitHub organization the badge should be `[![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/ropensci/pkgname?branch=master&svg=true)](https://ci.appveyor.com/project/individualaccount/pkgname)`.
- [ ] We're starting to roll out software metadata files to all ropensci packages via the Codemeta initiative, see https://github.com/ropensci/codemetar/#codemetar for how to include it in your package, after installing the package - should be easy as running codemetar::write_codemeta() in the root of your package.
- [ ] Activate Zenodo watching the repo
- [ ] Tag and create a release so as to create a Zenodo version and DOI
- [ ] Submit to JOSS using the Zenodo DOI.  We will tag it for expedited review.

Should you want to acknowledge your reviewers in your package DESCRIPTION, you can do so by making them `"rev"`-type contributors in the `Authors@R` field (with their consent).  More info on this [here](https://ropensci.github.io/dev_guide/building.html#authorship).

Welcome aboard! We'd love to host a blog post about your package - either a [short introduction to it with one example](https://ropensci.org/tech-notes/) or [a longer post with some narrative about its development or something you learned, and an example of its use](https://ropensci.org/blog/). If you are interested, review [the instructions](https://github.com/ropensci/roweb2#contributing-a-blog-post), and tag @stefaniebutland in your reply. She will get in touch about timing and can answer any questions.

We've started putting together a gitbook with our best practice and tips, [this chapter](https://ropensci.github.io/dev_guide/collaboration.html) starts the 3d section that's about guidance for after onboarding. Please tell us what could be improved, the corresponding repo is [here](https://github.com/ropensci/dev_guide).

**Comment by ShixiangWang** [comment]
@melvidoni I got a message when I transferred the repository

> You don’t have the permission to create repositories on ropensci

**Comment by melvidoni** [comment]
@ShixiangWang I sent you an invitation to be part of rOpenSci organisation in GItHub. If you accept it, you should be able to transfer the repo. (Transfer, do not create!).

**Comment by ShixiangWang** [comment]
@melvidoni Thanks, it works. 
There is a problem when I do this step `Activate Zenodo watching the repo`, I can't find my package in the Github list

![image](https://user-images.githubusercontent.com/25057508/62270085-abe42e80-b467-11e9-934d-aa7612881999.png)

It seems I should have rights to access the repositories of ropensci.

![image](https://user-images.githubusercontent.com/25057508/62270197-febde600-b467-11e9-91f3-f0aa98240b92.png)


I don't know where the problem comes from, the Zenodo (I just opened [an issue](https://github.com/zenodo/zenodo/issues/1814)) or the setting from ropensci? Please help me .

I also found I am in a team `auk`, please remove me from it.

![image](https://user-images.githubusercontent.com/25057508/62270311-4b092600-b468-11e9-8a0f-54d1ac52ee6a.png)

**Comment by melvidoni** [comment]
I removed you from that team @ShixiangWang. Let me try some things regarding Zenodo!

**Comment by melvidoni** [comment]
I removed you from that team @ShixiangWang, and added you to the package team. Can you accept that? (it shows as pending). Also, can you try adding Zenodo now?

**Comment by ShixiangWang** [comment]
Strange, I was already in the package team yesterday, so why is it showing pending?. I will try Zenodo as soon as possible.

**Comment by ShixiangWang** [comment]
@melvidoni Thanks, I can find the package in Zenodo now. :)

![image](https://user-images.githubusercontent.com/25057508/62337024-7857e180-b505-11e9-9a8e-5709ee6bc04b.png)

**Comment by melvidoni** [comment]
Awesome! Then, if you can tackle everything and then let me know, that would be great.

**Comment by ShixiangWang** [comment]
@melvidoni I have done everything in your todo list.

I have submitted URL https://doi.org/10.5281/zenodo.3358530 to JOSS.

If the DOI not found, please see https://zenodo.org/record/3358530

@stefaniebutland I would like to write a short introduction to this package.

Thanks all.
Shixiang

**Comment by stefaniebutland** [comment]
Good to hear that @ShixiangWang 

@melvidoni is there an element of the package, it's development, or applications that you think could be highlighted?

For a short introduction, a Tech Note is most appropriate. [Examples](https://ropensci.org/technotes/). It's important not to repeat information that's in the vignette; refer to it where needed. You might illustrate a compelling use case for the package.

Editorial and technical information about preparing your submission, including a template: https://github.com/ropensci/roweb2#contributing-a-blog-post. In the template YAML, for a Tech Note, set `categories: technotes`

Please suggest a deadline for submitting your draft. Don't hesitate to ask me questions. I'll be away Mon Aug 5. 

Thank you for doing this extra work. I hope the post will get more eyes on your work.

**Comment by ShixiangWang** [comment]
@stefaniebutland Thanks, can I submit the draft within a month? 

@melvidoni It seems that I should submit repository URL instead of Zenodo DOI to JOSS, please see https://github.com/openjournals/joss-reviews/issues/1622#issuecomment-517802770

**Comment by melvidoni** [comment]
@ShixiangWang That is weird! If they need URL repo, then that's it. Send the repo. I will advise editors to fix this on the approval comment template as well!

**Comment by ShixiangWang** [comment]
@melvidoni Yes 😭. Is there anything more I should do?

**Comment by melvidoni** [comment]
Not for now! Just send them the URL repo then.

**Comment by ShixiangWang** [comment]
@melvidoni The package has published on JOSS <https://joss.theoj.org/papers/10.21105/joss.01627>. Thanks for your work, the process is really fast.

**Comment by melvidoni** [comment]
@ShixiangWang That is great! If you don't mind, I'm tweeting your package, and closing the thread. Congrats and thank you for all the hard work!

**Comment by stefaniebutland** [comment]
> can I submit the draft within a month?

Yes, sounds good. I'll mark my calendar to check with you on Aug 27. Thank you.
